pkg/cloud/ibmcloud/actuators/cluster/actuator.go

```diff
@@ -17,38 +17,70 @@ limitations under the License.
 package cluster
 
 import (
+	"context"
+	"fmt"
+	"github.com/pkg/errors"
 	"k8s.io/klog"
+
+	providerv1 "sigs.k8s.io/cluster-api-provider-ibmcloud/pkg/apis/ibmcloud/v1alpha1"
+	"sigs.k8s.io/cluster-api-provider-ibmcloud/pkg/cloud/ibmcloud"
 	clusterv1 "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
-	client "sigs.k8s.io/cluster-api/pkg/client/clientset_generated/clientset/typed/cluster/v1alpha1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-// Actuator is responsible for performing cluster reconciliation
-type Actuator struct {
-	clustersGetter client.ClustersGetter
-}
-
-// ActuatorParams holds parameter information for Actuator
-type ActuatorParams struct {
-	ClustersGetter client.ClustersGetter
+type IBMCloudClient struct {
+	params ibmcloud.ActuatorParams
+	client client.Client
 }
 
 // NewActuator creates a new Actuator
-func NewActuator(params ActuatorParams) (*Actuator, error) {
-	return &Actuator{
-		clustersGetter: params.ClustersGetter,
+// NewActuator creates a new Actuator
+func NewActuator(params ibmcloud.ActuatorParams) (*IBMCloudClient, error) {
+	return &IBMCloudClient{
+		params: params,
+		client: params.Client,
 	}, nil
 }
 
 // Reconcile reconciles a cluster and is invoked by the Cluster Controller
-func (a *Actuator) Reconcile(cluster *clusterv1.Cluster) error {
+func (a *IBMCloudClient) Reconcile(cluster *clusterv1.Cluster) error {
+	if cluster == nil {
+		return fmt.Errorf("The cluster is nil, check your cluster configuration")
+	}
+
 	klog.Infof("Reconciling cluster %v.", cluster.Name)
 
-	klog.Infof("Reconciling cluster %v: Not implemented yet.", cluster.Name)
+	// Load provider status.
+	status, err := providerv1.ClusterStatusFromProviderStatus(cluster.Status.ProviderStatus)
+	if err != nil {
+		return errors.Errorf("failed to load cluster provider status: %v", err)
+	}
+
+	defer func() {
+		if err := a.storeClusterStatus(cluster, status); err != nil {
+			klog.Errorf("failed to store provider status for cluster %q in namespace %q: %v", cluster.Name, cluster.Namespace, err)
+		}
+	}()
+	return nil
+}
+
+func (a *IBMCloudClient) storeClusterStatus(cluster *clusterv1.Cluster, status *providerv1.IBMCloudClusterProviderStatus) error {
+	ext, err := providerv1.EncodeClusterStatus(status)
+	if err != nil {
+		return fmt.Errorf("failed to update cluster status for cluster %q in namespace %q: %v", cluster.Name, cluster.Namespace, err)
+	}
+	cluster.Status.ProviderStatus = ext
+
+	statusClient := a.params.Client.Status()
+	if err := statusClient.Update(context.Background(), cluster); err != nil {
```

Comment:
What about `context.TODO()`?

> Background returns a non-nil, empty Context. It is never canceled, has no 
> values, and has no deadline. It is typically used by the main function,  
> initialization, and tests, and as the top-level Context for incoming
> requests.

Reply:
it's copied from openstack side, I can change that